NMS-13792: Upgrade to Antora 3.1.1

indigo423 opened the pull request:
This PR migrates from Antora 2.3.4 to 3.1.1. With Antora 3 there is now an extension framework and the Lunr plugin is migrated and is now implemented as an extension. The Lunr extension got some fundamental rewrite and it required us to rebuild our search integration in our Antora UI

## Reviewers hint
We did a modification in the search.js to highlight search results in NMS-13540 which got busted. To get it back to work, I had to use the original file from their repository. For sustainability, we probably should look into getting the contribution [upstream](https://gitlab.com/antora/antora-lunr-extension) instead of maintaining it as a fork.

The customizations for admonitions, external Javascript for MathJAX are all moved now to the Antora UI and we don't need to maintain it per repository, i.e. cloud docs and docs.opennms.com.

The supplemental_ui directory needs to be renamed to match the Antora Lunr extension, which expects the directory to be existent. Additionally, the supplemental UI components for Admonitions and some CSS updates are moved to the Antora UI to be globally available instead of per-project.

The [site.zip](https://output.circle-artifacts.com/output/job/6098c568-d784-4310-a333-0361038d5249/artifacts/0/site.zip) can be used for local testing.

## Risk

If we break it, we can revert this PR and fix it :)

JIRA: https://issues.opennms.org/browse/NMS-13792

Bonrob2 reviewed (commented):
makes sense to me, but unable to test it.

indigo423 commented:
> makes sense to me, but unable to test it.

yes you can test it 

1. Download the ZIP file from the last build, like here: https://1666-251686651-gh.circle-artifacts.com/0/site.zip
2. Extract it
3. Open the index.html in your browser

Report issues if you find some :)

indigo423 commented:
Instead of 29 you see now "latest" as part of the URL for example.

RangerRick commented:
So I assume before we merge this, we need to update the redirects to send you to `latest` instead of the short version.

indigo423 commented:
> So I assume before we merge this, we need to update the redirects to send you to `latest` instead of the short version.

yes, before or after the merge

Bonrob2 commented:
> > makes sense to me, but unable to test it.
> 
> yes you can test it
> 
>     1. Download the ZIP file from the last build, like here: https://1666-251686651-gh.circle-artifacts.com/0/site.zip
> 
>     2. Extract it
> 
>     3. Open the index.html in your browser
> 
> 
> Report issues if you find some :)

Looks good to me! It will be great to cross-reference between docs and components.

indigo423 commented:
@GanimanSwift Just pulling you in here for this PR which introduces the "latest" floating tag instead of the version numbers "29" for Horizon and "2021". Do you think this is a showstopper? We would need to adjust our redirect regexp. If you think there is too much risk with it, I can easily remove it and make a dedicated PR with it. What do you think?

Bonrob2 commented:
@GanimanSwift are you able to address @indigo423's question? I'd like to merge this if possible.

RangerRick commented:
Ah so if it does "latest" it doesn't also do one based on the version?

indigo423 commented:
> Ah so if it does "latest" it doesn't also do one based on the version?

Yes for Horizon it would only be only one version, right now we could show 29.0.6 and the URL path would have `latest` instead of 29.0.6, 29, 30 or whatever our last release is.

For Meridian, latest would point to the most current Meridian 2021 release. All the older versions have the version number in the path.

For all other projects, it would be the same. There are two exceptions, ALEC and OpenNMS-JS, we publish SNAPSHOT releases. They would get the `latest` floating tag. We would need to remove these SNAPSHOT docs to have `latest` to be consistently the last release we want to encourage people to run and to upgrade to.

fooker reviewed (dismissed):
I can try to bring it back/bring it upstream. Should we make this a separate issue or is it a stopper for this?

indigo423 commented:
> I can try to bring it back/bring it upstream. Should we make this a separate issue or is it a stopper for this?

I don't think we should make it a show stopper for this PR. We can reopen NMS-13540 and track it there?

Bonrob2 commented:
@indigo423 are we able to merge this now? Chris has finished the work on his side.

RangerRick reviewed (dismissed):
yeah, I like not doing the `latest` on the antora side, makes more sense to do it with a symlink as part of the deploy

Bonrob2 commented:
@indigo423 can we merge this now, or do we need to do additional work?

Bonrob2 commented:
@indigo423 or @RangerRick ... doing some clean up and noticed this is still sitting around. Could we move forward and upgrade Antora?

Bonrob2 commented:
I have re-opened NMS-13540 to request the Lunr highlighting work be done upstream. IT has done the regex work. Could we merge this PR? @indigo423?

indigo423 committed:
Update to Antora 3.1.1

Update repository to use Antora 3.1 instead of 2.3.4. Reworked the Lunr to use introduced extension framework in
Antora 3.x. Migrated the new Antora UI 3.1.0 which uses the project scope search feature.

indigo423 commented on antora-playbook.yml:
```diff
@@ -80,9 +86,10 @@ content:
   bundle:
-    url: https://github.com/opennms-forge/antora-ui-opennms/releases/download/v3.0.1/ui-bundle.zip
+    #url: https://github.com/opennms-forge/antora-ui-opennms/releases/download/v3.0.1/ui-bundle.zip
+    url: https://output.circle-artifacts.com/output/job/b282354c-4796-42e9-be8f-7a88bc920ed3/artifacts/0/ui-bundle.zip
```
We have a dependency on the Antora UI update here: https://github.com/opennms-forge/antora-ui-opennms/pull/27

⚠️ This is a placeholder here that links to a CircleCI build artifact for testing.

indigo423 commented:
Released a new version with the Antora UI and used it here in the repository instead of the build artifact.

Bonrob2 reviewed (commented):
LGTM, but will defer to Ben for approval.

RangerRick reviewed (approved):
assuming the folks who are affected by it (product? support?) are ok with search hilighting going away, I'm OK with this
